packages/core/src/zip/__tests__/select.js

```diff
@@ -0,0 +1,238 @@
+'use strict'
+
+import I from 'immutable'
+import R from 'ramda'
+
+import { elementZipper } from '../'
+import { childrenProperty, isElement, isOfKind } from '../../data'
+
+import {
+  isStringArray,
+  isLocationArray,
+  child,
+  ancestors,
+  descendants,
+  children,
+  propEq,
+  select,
+  ofKind,
+} from '../select'
+
+const zipper = data => elementZipper({})(data)
+
+const data = {
+  kind: ['navigation', 'tabs'],
+  [childrenProperty]: ['tabs', 'settings'],
+  title: 'App',
+  tabs: [
+    {
+      kind: ['tab'],
+      title: 'Murder Bot',
+    },
+    {
+      kind: ['tab'],
+      title: 'Dr Mensah',
+      [childrenProperty]: ['elements', 'deeplink'],
+      elements: [
+        {
+          kind: ['element'],
+          title: 'robot',
+        },
+        {
+          kind: ['element'],
+          title: 'human',
+        },
+      ],
+      deeplink: {
+        kind: ['link'],
+        title: 'goodreads',
+      },
+    },
+  ],
+  settings: {
+    kind: ['settings'],
+    title: 'Martha Wells',
+  },
+}
+
+describe('zip.select', () => {
+  describe('helpers', () => {
+    it('isStringArray', () => {
+      expect(isStringArray(['string', 'test'])).toEqual(true)
+      expect(isStringArray(['string', null])).toEqual(false)
+      expect(isStringArray(['string', undefined])).toEqual(false)
+      expect(isStringArray([{}, 'test'])).toEqual(false)
+      expect(isStringArray(['string', []])).toEqual(false)
+      expect(isStringArray([[]])).toEqual(false)
+      expect(isStringArray([])).toEqual(true)
+      expect(isStringArray(null)).toEqual(false)
+      expect(isStringArray(undefined)).toEqual(false)
+    })
+    it('isLocationArray', () => {
+      expect(isLocationArray()).toEqual(false)
+      expect(isLocationArray(null)).toEqual(false)
+      expect(isLocationArray(undefined)).toEqual(false)
+      expect(isLocationArray([])).toEqual(false)
+      expect(isLocationArray(['test'])).toEqual(false)
+      expect(isLocationArray([zipper(I.fromJS(data))])).toEqual(true)
+      expect(
+        isLocationArray([zipper(I.fromJS(data)), zipper(I.fromJS(data))])
+      ).toEqual(true)
+      expect(
+        isLocationArray([
+          zipper(I.fromJS(data)),
+          'Not a Zipper',
+          zipper(I.fromJS(data)),
+        ])
+      ).toEqual(false)
+    })
+  })
+
+  describe('predicates', () => {
+    it('child', () => {
```

Comment:
This is a **motion** fn. A **motion** fn has the signature `(...args, node) -> node`. I.e. will take a number of arguments (that determine its behavior, a node at the end and will return a node which is the result of the motion).

Reply:
Extracted motions.